### ISSUE TITLE
Mob - AI LoS Breaking / Pull (Caster Mob) (BB #592)

### ISSUE THREAD

**Issue by OregonAsari** [OregonCore/OregonCore#592]
This issue was migrated from bitbucket.
**Original Reporter:** Areradon
**Original Date:** 18.05.2014 08:55:31 GMT+0000
**Original Priority:** major
**Original Type:** bug
**Original State:** resolved
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592

<hr>

How works: When you pull a Mob that should only cast and run Out of Sight the Mob only stand there.
How should work (with proofs): When you break LoS the mob should stop cast and run to the player that the NPC is again in LoS to cast spells.
https://www.youtube.com/watch?v=gXn7HRgYe1c

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 04.08.2014 04:32:51 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11585243

<hr>

```
#!c++

# HG changeset patch
# User LordUsagi
# Date 1393743924 14400
#      Sun Mar 02 03:05:24 2014 -0400
# Node ID 7cc5c157bbdd46bfe646df8709b169332451dc6e
# Parent  20fbebd8463281231351798cc31b4a79b0ffa673
[Core/CreatureAI] EventAI creatures will no longer stand around when target is
                  out of LOS and will proceed to chase the target down.

diff -r 20fbebd84632 -r 7cc5c157bbdd src/game/CreatureEventAI.cpp
--- a/src/game/CreatureEventAI.cpp  Sun Mar 02 02:59:33 2014 -0400
+++ b/src/game/CreatureEventAI.cpp  Sun Mar 02 03:05:24 2014 -0400
@@ -482,6 +482,9 @@

             //Allowed to cast only if not casting (unless we interrupt ourself) or if spell is triggered
             bool canCast = !caster->IsNonMeleeSpellCasted(false) || (action.cast.castFlags & (CAST_TRIGGERED | CAST_INTURRUPT_PREVIOUS));
+            
+            //Determine if creature can reach the target
+            bool canReach = caster->IsWithinLOSInMap(target) && caster->IsWithinLOS(target->GetPositionX(), target->GetPositionY(), target->GetPositionZ());

             // If cast flag CAST_AURA_NOT_PRESENT is active, check if target already has aura on them
             if (action.cast.castFlags & CAST_AURA_NOT_PRESENT)
@@ -516,6 +519,17 @@
                     }
                     else
                     {
+                        if(!canReach)
+                        {
+                            if (me->GetMotionMaster()->GetCurrentMovementGeneratorType() == TARGETED_MOTION_TYPE || me->GetMotionMaster()->GetCurrentMovementGeneratorType() == ASSISTANCE_MOTION_TYPE || me->GetMotionMaster()->GetCurrentMovementGeneratorType() == IDLE_MOTION_TYPE)
+                            {
+                                AttackDistance = 0.0f;
+                                AttackAngle = 0.0f;
+
+                                me->GetMotionMaster()->MoveChase(me->getVictim(), AttackDistance, AttackAngle);
+                            }
+                        }
+
                         //Interrupt any previous spell
                         if (caster->IsNonMeleeSpellCasted(false) && action.cast.castFlags & CAST_INTURRUPT_PREVIOUS)
                             caster->InterruptNonMeleeSpells(false);

```

The above patch needs testing. Anyone of our wonderful Oregon users could be kind enough to test :

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Areradon
**Original Date:** 24.08.2014 22:59:29 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11960996

<hr>

I have now played long on a wotlk server and i have lookes a bit into the Smart Ai from Trinity Core. These works much better than the CreatureAI from OregonCore. As an example you can see the Bleeding Hollow Archer (http://www.wowhead.com/npc=17270) which works like he should be. (Switching between Range mode and Melee mode)
https://www.youtube.com/watch?v=gXn7HRgYe1c

"Comment table in smart_ai from TDB 3.3.5"
- Bleeding Hollow Archer - Out of Combat - Disable Combat Movement
- Bleeding Hollow Archer - Out of Combat - Stop Attacking
- Bleeding Hollow Archer - On Aggro - Cast 'Multi-Shot'
- Bleeding Hollow Archer - On Aggro - Increment Phase By 1
- Bleeding Hollow Archer - Within 5-30 Range - Cast 'Multi-Shot'
- Bleeding Hollow Archer - Within 5-30 Range - Set Sheath Ranged
- Bleeding Hollow Archer - On Aggro - Cast 'Multi-Shot'
- Bleeding Hollow Archer - On Aggro - Increment Phase By 1
- Bleeding Hollow Archer - Within 5-30 Range - Cast 'Multi-Shot'
- Bleeding Hollow Archer - Within 5-30 Range - Set Sheath Ranged
- Bleeding Hollow Archer - Within 25-80 Range - Enable Combat Movement
- Bleeding Hollow Archer - Within 25-80 Range - Start Attacking
- Bleeding Hollow Archer - Within 0-5 Range - Enable Combat Movement
- Bleeding Hollow Archer - Within 0-5 Range - Set Sheath Melee
- Bleeding Hollow Archer - Within 0-5 Range - Start Attacking
- Bleeding Hollow Archer - Within 5-15 Range - Disable Combat Movement
- Bleeding Hollow Archer - Within 5-15 Range - Stop Attacking
- Bleeding Hollow Archer - In Combat - Cast 'Aimed Shot'
- Bleeding Hollow Archer - In Combat - Set Sheath Ranged
- Bleeding Hollow Archer - On Evade - Set Sheath Melee

I dont know if you Patch can also fix this. But i now prefer the Smart Ai from TC-Wotlk since i worked with her for 2 month.. (;

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 25.08.2014 03:42:06 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11963245

<hr>

SmartAI is definitely something that I'll be bringing to OregonCore, seeing as when we originally started out we were the 2.4.3 continuation of TrinityCore, going forward I think that's something I personally want to stick true to.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 25.08.2014 04:54:41 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11963875

<hr>

The only issue in bringing SmartAI to OregonCore at the moment is the lack of DB Developers. If I knew there was more interest in fixing issues in quests, dungeons etc. using SmartAI I would be into implementing it quicker.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Areradon
**Original Date:** 25.08.2014 07:48:59 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11966139

<hr>

With SmartAI its defnitly much easier to fix Dungeons. I would pref to have an working SmartAI before have working quests. :D

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 25.08.2014 07:50:22 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11966173

<hr>

But we'd need smartai sql developers.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Areradon
**Original Date:** 25.08.2014 08:27:44 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11967171

<hr>

The Most things you need, are done from tc2.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 25.08.2014 11:47:00 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11970660

<hr>

Not exactly, I think their DB has changed quite a bit, I'm not sure if it uses IDs or GUIDs, if the latter, we have an issue, if the prior then we're okay.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Areradon
**Original Date:** 25.08.2014 20:05:20 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-11980256

<hr>

It use both.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Areradon
**Original Date:** 30.10.2014 22:36:28 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-13329583

<hr>

How is this going so far?

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 30.10.2014 22:40:21 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-13329620

<hr>

Awaiting Unix developers /testers on dbdev branch. Considering another
method.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Areradon
**Original Date:** 23.01.2015 13:09:25 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-15037054

<hr>

Which method are you testing? I startet to implement a the SmartAi from TC into my personal oregoncore. its realy hard but maybe i get it to work.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Areradon
**Original Date:** 19.02.2015 11:25:05 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-15795006

<hr>

After a long time i now tested the Patch from you.

This dont fix it.

```
#!c++

LoS

diff -r 27dbc645149e -r 6484a59a3c58 src/game/CreatureEventAI.cpp
--- a/src/game/CreatureEventAI.cpp  Sat Jan 24 23:41:13 2015 +0100
+++ b/src/game/CreatureEventAI.cpp  Thu Feb 19 12:19:30 2015 +0100
@@ -1409,6 +1409,9 @@
     if (!Triggered && me->HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_SILENCED))
         return false;

+   //Target is not within los with caster 
+       if (!m_creature->IsWithinLOSInMap(Target)) return false;
+
     //Check for power
     if (!Triggered && me->GetPower((Powers)Spell->powerType) < SpellMgr::CalculatePowerCost(Spell, me, SpellMgr::GetSpellSchoolMask(Spell)))
         return false;
```

This Check is enough for LoS. But after getting LoS again they chase you and dont cast again.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** desteny
**Original Date:** 23.02.2015 17:32:51 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-15869133

<hr>

If you look at CreatureEventAI::ProcessAction in case ACTION_T_CAST

you will see in line 517, the movemnet generator is changing AttackDistance
this change will not be reverted if you are again in LOS.

This needs to be rewritten

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Areradon
**Original Date:** 23.02.2015 21:16:08 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-15873573

<hr>

```
#!c++

case SMART_ACTION_CAST:
{
ObjectList* targets = GetTargets(e, unit);
if (!targets)
break;
for (ObjectList::const_iterator itr = targets->begin(); itr != targets->end(); ++itr)
{
if (!IsUnit(*itr))
continue;
if (!(e.action.cast.flags & SMARTCAST_AURA_NOT_PRESENT) || !(*itr)->ToUnit()->HasAura(e.action.cast.spell))
{
if (me)
{
if (e.action.cast.flags & SMARTCAST_INTERRUPT_PREVIOUS)
me->InterruptNonMeleeSpells(false);
if (e.action.cast.flags & SMARTCAST_COMBAT_MOVE)
{
// If cast flag SMARTCAST_COMBAT_MOVE is set combat movement will not be allowed
// unless target is outside spell range, out of mana, or LOS.
bool _allowMove = false;
SpellInfo const* spellInfo = sSpellMgr->EnsureSpellInfo(e.action.cast.spell);
int32 mana = me->GetPower(POWER_MANA);
if (me->GetDistance(*itr) > spellInfo->GetMaxRange(true) ||
me->GetDistance(*itr) < spellInfo->GetMinRange(true) ||
!me->IsWithinLOSInMap(*itr) ||
mana < spellInfo->CalcPowerCost(me, spellInfo->GetSchoolMask()))
_allowMove = true;
ENSURE_AI(SmartAI, me->AI())->SetCombatMove(_allowMove);
}
me->CastSpell((*itr)->ToUnit(), e.action.cast.spell, (e.action.cast.flags & SMARTCAST_TRIGGERED) != 0);
}
else if (go)
go->CastSpell((*itr)->ToUnit(), e.action.cast.spell, (e.action.cast.flags & SMARTCAST_TRIGGERED) != 0);
TC_LOG_DEBUG("scripts.ai", "SmartScript::ProcessAction:: SMART_ACTION_CAST:: %s: %u casts spell %u on target %u with castflags %u",
(me ? me->GetGUID() : go->GetGUID()).GetTypeName(), me ? me->GetGUIDLow() : go->GetGUIDLow(), e.action.cast.spell, (*itr)->GetGUIDLow(), e.action.cast.flags);
}
else
TC_LOG_DEBUG("scripts.ai", "Spell %u not cast because it has flag SMARTCAST_AURA_NOT_PRESENT and the target (%s) already has the aura", e.action.cast.spell, (*itr)->GetGUID().ToString().c_str());
}
delete targets;
break;
}
```

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 03.06.2015 15:20:55 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/592/#comment-18567352

<hr>

Resolved in 4d6ad6af3757604183aeaf89ccda012fe8616f72
